core-primitives/node-api/metadata-provider/src/lib.rs

```diff
@@ -72,46 +69,14 @@ where
 
 impl<NodeMetadata> AccessNodeMetadata for NodeMetadataRepository<NodeMetadata>
 where
-	NodeMetadata:,
+	NodeMetadata: NodeMetadataTrait + Clone,
 {
 	type MetadataType = NodeMetadata;
 
-	fn get_from_metadata<F, R>(&self, getter_function: F) -> Result<R>
-	where
-		F: FnOnce(&Self::MetadataType) -> R,
-	{
+	fn get(&self) -> Result<Self::MetadataType> {
 		match self.metadata_lock.read().expect("Lock poisoning").deref() {
-			Some(metadata) => Ok(getter_function(metadata)),
+			Some(metadata) => Ok(metadata.clone()),
```

Comment:
No problem at all.

Yea I also oversaw the fact that the `get()` can be frequently called and metadata clone is costly. Getting a simple reference in this case is not so easy as `RwLockReadGuard` is dropped when out of scope.

Let's close this PR until we find a better way :)

Reply:
One thing: do you consider the change in [`core-primitives/node-api/metadata/src/lib.rs`](https://github.com/integritee-network/worker/pull/1148/files#diff-d8c60e42efe0918ac6d69ed3f5df53be4502388a234c46c52baf5b56f6fd7a49) valuable? With the `NodeMetadataTrait` we have the benefits that we only need to add more bounds in one place instead of modifying code across multiple files. We have 5 bounds already 😂 

If you think so I can adjust the PR to solely incorporate this change.

Reply:
For me this is fine, but the way you would usually do this is:

```rust

trait NodeMetadata: TeerexCallIndexes + SidechainCallIndexes {}
impl<T: TeerexCallIndexes + SidechainCallIndexes> for T {}

```

So we don't have to re-implement it for the mock.